### PR TITLE
[Dataset] Bump `numpy >=1.20` dependency

### LINKED ISSUE #20258: [release][core-nightly] inference is failing.
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Ray Core

### What happened + What you expected to happen

https://beta.anyscale.com/o/anyscale-internal/projects/prj_2xR6uT6t7jJuu1aCwWMsle/clusters/ses_nyHPVp7fNykxK9vAV8UCVLrn?command-history-section=command_history

### Versions / Dependencies

N/A

### Reproduction script

N/A

### Anything else

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by wuisawesome**
## Why are these changes needed?

We need to ensure our numpy version is greater than 1.20. In 1.19 and under, the following code doesn't work

```
import pandas as pd
import torch
pd.DataFrame([torch.arange(27).reshape(3,3,3) for _ in range(10)])
```

Right now, this is pretty fundamental to how we build datasets of tensors, which is common when doing last-mile data preprocessing (i.e. a preprocessed image is most likely represented as a tensor). 

Note that 1.20 dropped support for python 3.6 so the best we can do is enforced this versioning when using newer python versions. 

## Related issue number

Closes #20258

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/setup.py [hunk 1]
```diff
@@ -219,6 +219,10 @@ def get_packages(self):
         ],
     }
 
+    if sys.version_info >= (3, 7):
+        # Numpy dropped python 3.6 support in 1.20.
+        setup_spec.extras["data"].append("numpy >= 1.20")
```
Comment (jjyao) [inline python/setup.py:224]: I'm wondering what's the difference between data extras and https://github.com/ray-project/ray/blob/master/python/requirements/data_processing/requirements_dataset.txt

Comment (wuisawesome) [reply python/setup.py:224]: Data extras is what's actually installed when you do `pip install ray[data]`. Those requirements files should be for additional ray developer add-ons.

**Comment by ericl** [comment]
@wuisawesome can you elaborate on what "doesn't work mean"? Can we raise a better error message in this condition?

**Comment by wuisawesome** [comment]
Yeah by "doesn't work" I mean it throws this error

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ubuntu/.local/lib/python3.8/site-packages/pandas/core/frame.py", line 711, in __init__
    mgr = ndarray_to_mgr(
  File "/home/ubuntu/.local/lib/python3.8/site-packages/pandas/core/internals/construction.py", line 304, in ndarray_to_mgr
    values = _prep_ndarray(values, copy=copy)
  File "/home/ubuntu/.local/lib/python3.8/site-packages/pandas/core/internals/construction.py", line 540, in _prep_ndarray
    values = np.array([convert(v) for v in values])
ValueError: only one element tensors can be converted to Python scalars
```

Note that other tensors (like ndarray) also don't work, but throw different error messages. To make better error messages we could try to swallow all errors and tell people to upgrade their numpy version (seems a little scary) or start inspecting the contents of the dataset and doing special case checks for the various tensor types if numpy < 1.20 on python 3.6.

**Review by ericl — APPROVED** [review]

**Comment by wuisawesome** [comment]
test_output is broken on master

**Pull request merged**